src/Sulu/Component/Content/Repository/ContentRepository.php

```diff
@@ -0,0 +1,636 @@
+     */
+    private function getLocalesByWebspaceKey($webspaceKey)
+    {
+        $webspace = $this->webspaceManager->findWebspaceByKey($webspaceKey);
+
+        return array_map(
+            function (Localization $localization) {
+                return $localization->getLocalization();
+            },
+            $webspace->getAllLocalizations()
+        );
+    }
+
+    /**
+     * Returns array of locales for webspaces.
+     *
+     * @return string[]
+     */
+    private function getLocales()
+    {
+        return array_map(
+            function (Localization $localization) {
+                return $localization->getLocalization();
+            },
+            $this->webspaceManager->getAllLocalizations()
+        );
+    }
+
+    /**
+     * Append mapping selects to given query-builder.
+     *
+     * @param QueryBuilder $queryBuilder
+     * @param MappingInterface $mapping Includes array of property names.
+     * @param string[] $locales
+     */
+    private function appendMapping(QueryBuilder $queryBuilder, MappingInterface $mapping, $locales)
+    {
+        $properties = $mapping->getProperties();
+        $properties[] = 'template';
+        $properties[] = 'shadow-on';
+        foreach ($properties as $propertyName) {
+            $this->appendSingleMapping($queryBuilder, $propertyName, $locales);
+        }
+    }
+
+    /**
+     * Append mapping selects for a single property to given query-builder.
+     *
+     * @param QueryBuilder $queryBuilder
+     * @param string $propertyName
+     * @param string[] $locales
+     */
+    private function appendSingleMapping(QueryBuilder $queryBuilder, $propertyName, $locales)
+    {
+        foreach ($locales as $locale) {
+            $alias = sprintf('%s%s', $locale, str_replace('-', '_', ucfirst($propertyName)));
+
+            $queryBuilder->addSelect(
+                'node',
+                $this->propertyEncoder->localizedContentName($propertyName, $locale),
+                $alias
+            );
+        }
+    }
+
+    /**
+     * Resolve a single result row to a content object.
+     *
+     * @param Row $row
+     * @param string $locale
+     * @param string $webspaceKey
+     * @param MappingInterface $mapping Includes array of property names.
+     * @param UserInterface $user
+     *
+     * @return Content
+     */
+    private function resolveContent(
+        Row $row,
+        $locale,
+        $webspaceKey,
+        MappingInterface $mapping,
+        UserInterface $user = null
+    ) {
+        if (!$webspaceKey) {
+            $webspaceKey = $this->nodeHelper->extractWebspaceFromPath($row->getPath());
+        }
+
+        $originalLocale = $locale;
+        $ghostLocale = $this->localizationFinder->findAvailableLocale(
+            $webspaceKey,
+            $this->resolveAvailableLocales($row),
+            $locale
+        );
+
+        $type = null;
+        if ($row->getValue('shadowOn')) {
+            if (!$mapping->hydrateShadow()) {
```

Comment:
`hydrateShadow` sounds like the method is doing something, but I guess it only returns a boolean. Can you come up with a better name?